linera-core/src/local_node.rs

```diff
@@ -203,26 +208,144 @@ where
+        .collect::<Vec<_>>())
+    }
+
+    // Given a list of missing `BlobId`s and a `Certificate` for a block:
+    // - Make sure they're required by the block provided
+    // - Make sure there's no duplicate blobs being requested
+    // - Searches for the blob in different places of the local node: blob cache,
+    //   chain manager's pending blobs, and blob storage.
+    pub async fn find_missing_blobs(
+        &self,
+        certificate: &Certificate,
+        missing_blob_ids: &Vec<BlobId>,
+        chain_id: ChainId,
+    ) -> Result<Vec<HashedBlob>, NodeError> {
+        if missing_blob_ids.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Find the missing blobs locally and retry.
+        let required = match certificate.value() {
+            CertificateValue::ConfirmedBlock { executed_block, .. }
+            | CertificateValue::ValidatedBlock { executed_block, .. } => {
+                executed_block.required_blob_ids()
+            }
+            CertificateValue::Timeout { .. } => HashSet::new(),
+        };
+        for blob_id in missing_blob_ids {
+            if !required.contains(blob_id) {
+                warn!(
+                    "validator requested blob {:?} but it is not required",
+                    blob_id
+                );
+                return Err(NodeError::InvalidChainInfoResponse);
+            }
+        }
+        let mut unique_missing_blob_ids = missing_blob_ids.iter().cloned().collect::<HashSet<_>>();
+        if missing_blob_ids.len() > unique_missing_blob_ids.len() {
+            warn!("blobs requested by validator contain duplicates");
+            return Err(NodeError::InvalidChainInfoResponse);
+        }
+
+        let (found_blobs, not_found_blobs): (HashMap<BlobId, HashedBlob>, Vec<BlobId>) = self
+            .recent_hashed_blobs()
+            .await
+            .try_get_many(missing_blob_ids.clone())
+            .await;
+        let found_blob_ids = found_blobs.clone().into_keys().collect::<HashSet<_>>();
+        let mut found_blobs = found_blobs.clone().into_values().collect::<Vec<_>>();
+
+        unique_missing_blob_ids = unique_missing_blob_ids
+            .difference(&found_blob_ids)
+            .copied()
+            .collect::<HashSet<BlobId>>();
+        let chain_manager_pending_blobs = self
+            .chain_state_view(chain_id)
+            .await?
+            .manager
+            .get()
+            .pending_blobs
+            .clone();
+        for blob_id in not_found_blobs {
+            if let Some(blob) = chain_manager_pending_blobs.get(&blob_id).cloned() {
+                found_blobs.push(blob);
+                unique_missing_blob_ids.remove(&blob_id);
+            }
+        }
+
+        let storage = self.storage_client();
+        found_blobs.extend(
+            future::join_all(
+                unique_missing_blob_ids
+                    .into_iter()
+                    .map(|blob_id| storage.read_hashed_blob(blob_id)),
```

Comment:
Ok, I can add that also, but I think we need both because for one blob it would either fail or return the blob. For multiple blobs, it'll have to be best effort i.e. return a `Vec<Option<HashedBlob>>`